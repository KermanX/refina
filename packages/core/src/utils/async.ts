import { Prelude } from "../constants";

/**
 * The state of an await call.
 */
type AwaitState =
  | {
      type: "pending";
    }
  | {
      type: "fulfilled";
      value: unknown;
    }
  | {
      type: "error";
      reason: unknown;
    };

Prelude.registerFunc(
  "await",
  function (
    ckey: string,
    executor: () => Promise<unknown>,
    id: string | number = "",
  ): boolean {
    const refTreeNode = this.$$currentRefTreeNode;

    if (!refTreeNode[ckey]) {
      // The await call is not started.
      refTreeNode[ckey] = {
        type: "pending",
      } as AwaitState;
      executor()
        .then(value => {
          refTreeNode[ckey] = {
            type: "fulfilled",
            value,
          } as AwaitState;
          this.$update();
        })
        .catch(reason => {
          refTreeNode[ckey] = {
            type: "error",
            reason,
          } as AwaitState;
          this.$update();
        });
    }

    const state = refTreeNode[ckey] as AwaitState;
    if (state.type === "pending") {
      return false;
    } else if (state.type === "fulfilled") {
      // @ts-ignore
      this[`$awaited${id}`] = state.value;
      return true;
    } else {
      throw state.reason;
    }
  },
);

declare module "../context/base" {
  interface ContextFuncs<C> {
    /**
     * Use data from an async call when rendering.
     *
     * **Note**: Use `try`/`catch` to handle errors.
     *
     * @example
     * ```ts
     * if (_.await(() => fetch("https://example.com"))) {
     *   // When the promise is fulfilled, _.await returns true.
     *
     *   _.p(_.$awaited.statusText);
     *
     *   // You can also use a custom id for a nesting await call.
     *   _.await(() => _.$awaited.text(), "Text") && _.p(_.$awaitedText);
     * } else {
     *   // When the promise is pending, _.await returns false.
     *
     *   _.p("Loading...");
     * }
     * ```
     *
     * @param executor The async function to execute. This function is called only for once.
     * @param id The id of the await call. If not provided, the id is an empty string. Access the awaited data with `_.$awaited<id>`.
     * @throws The error thrown by the async function.
     */
    await: (<T, N extends string | number>(
      executor: () => Promise<T>,
      id: N,
    ) => // @ts-ignore
    this is Record<`$awaited${N}`, T>) &
      (<T>(executor: () => Promise<T>) => // @ts-ignore
      this is {
        $awaited: T;
      });
  }
}
