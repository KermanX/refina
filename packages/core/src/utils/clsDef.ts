import { Context } from "../context";

/**
 * Create a class name string.
 *
 * @param cls The class name.
 */
export function $clsStr<const S extends string>(cls: S): S;
/**
 * Create a class name string.
 *
 * @param template The template string of the class name.
 * @param args The arguments of the template string.
 */
export function $clsStr(
  template: TemplateStringsArray,
  ...args: unknown[]
): string;
export function $clsStr(s: string | TemplateStringsArray, ...args: unknown[]) {
  if (typeof s === "string") {
    return s;
  }
  return String.raw(s, ...args);
}

/**
 * Create a function that applies the given class name to the context.
 *
 * @param cls The class name.
 */
export function $clsFunc(cls: string): (_: Context) => true;
/**
 * Create a function that applies the given class name to the context.
 *
 * @param template The template string of the class name.
 * @param args The arguments of the template string.
 */
export function $clsFunc(
  template: TemplateStringsArray,
  ...args: unknown[]
): (_: Context) => true;
export function $clsFunc(...args: unknown[]) {
  // @ts-ignore
  return (_: Context) => _.$cls(...args);
}
