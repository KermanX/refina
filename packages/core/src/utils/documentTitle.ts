import { Prelude } from "../constants";
import { D, getD } from "../data";

Prelude.registerFunc(
  "documentTitle",
  function (_ckey: string, title: D<string>): boolean {
    if (this.$updateState) {
      document.title = getD(title);
    }
    return true;
  },
);

declare module "../context/base" {
  interface ContextFuncs<C> {
    /**
     * Set the document title.
     *
     * **Warning**: The document title will be overwritten by the last call to this function.
     */
    documentTitle: never extends C["enabled"]
      ? (title: D<string>) => void
      : never;
  }
}
