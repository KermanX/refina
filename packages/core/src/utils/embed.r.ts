import { OutputComponent } from "../component";
import { Prelude } from "../constants";
import { Context, IntrinsicContext } from "../context";
import { D, getD } from "../data";
import { Content } from "../dom";
import { View } from "../view";

const contentCache = new Map<string, Content<any>>();

type EmbededContent<Args extends any[]> =
  | Content<Args>
  | (() => Promise<{
      default: Content<Args>;
    }>);

@Prelude.outputComponent("embed")
export class Embed extends OutputComponent {
  main<Args extends any[]>(_: Context, content: D<EmbededContent<Args>>, ...args: Args): void {
    const contentValue: EmbededContent<Args> = contentCache.get(this.$ikey) ?? getD(content);

    const context = new IntrinsicContext(_.$app);

    if (typeof contentValue === "function") {
      const ret: unknown = contentValue(context as unknown as Context, ...args);
      if (ret instanceof Promise) {
        _.t`Loading module...`;
        ret.then((r) => {
          contentCache.set(this.$ikey, r.default);
          _.$app.update();
        });
      } else {
        _.$firstDOMNode = context.$firstDOMNode;
        _.$firstHTMLELement = context.$firstHTMLELement;
      }
    } else {
      _.t(String(contentValue));
    }
  }
}

declare module "../context" {
  interface CustomContext<C> {
    embed: never extends C
      ? <Args extends any[]>(
          content: D<
            | Content<Args>
            | (() => Promise<{
                default: Content<Args>;
              }>)
          >,
          ...args: Args
        ) => void
      : never;
  }
}
