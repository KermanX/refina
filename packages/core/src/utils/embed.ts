import { Prelude } from "../constants";
import { D, getD } from "../data";
import { Content } from "../dom";

declare module "../component" {
  interface Components {
    /**
     * Embed a content into the current output component.
     *
     * **Warning**: **DO NOT** call view functions directly in your code,
     *  unless you are **very** sure that the view will only be called once in your view / component.
     *  Use this function instead.
     *
     * @param content The content to embed.
     * @param args The arguments to pass to the content, if the content is a view function.
     */
    embed<Args extends any[]>(content: D<Content<Args>>, ...args: Args): void;
  }
}

Prelude.outputComponents.embed = function (_) {
  return (content, ...args) => {
    const contentValue = getD(content);
    if (typeof contentValue === "function") {
      try {
        contentValue(_, ...args);
        if (import.meta.env.DEV) {
          _.$lowlevel.$$assertEmpty();
        }
      } catch (e) {
        this.$app.callHook("onError", e);
      }
    } else {
      _.$lowlevel.$$t("_t", contentValue);
    }
  };
};

/**
 * A function that returns a promise of a object whose default export is a content.
 *
 * @example
 * ```ts
 * const loader: AsyncContentLoader = () => import("./myView.ts");
 * ```
 */
export type AsyncContentLoader<Args extends any[]> = () => Promise<{
  default: Content<Args>;
}>;

declare module "../component" {
  interface Components {
    /**
     * Like `embed`, but the content is loaded asynchronously.
     *
     * @example
     * ```ts
     * _.asyncEmbed(() => import("./myView.ts"));
     * ```
     *
     * @param contentLoader A function that returns a promise of a object whose default export is a content.
     * @param args The arguments to pass to the content, if the content is a view function.
     */
    asyncEmbed<Args extends any[]>(
      contentLoader: D<AsyncContentLoader<Args>>,
      ...args: Args
    ): void;
  }
}

Prelude.outputComponents.asyncEmbed = function (_) {
  let loadedContent: Content<any> | null = null;
  return (contentLoader, ...args) => {
    if (loadedContent) {
      _.embed(loadedContent, ...args);
    } else {
      getD(contentLoader)().then(v => {
        loadedContent = v.default;
        _.$update();
      });
    }
  };
};
