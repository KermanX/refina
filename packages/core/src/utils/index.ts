export * from "./async";
export * from "./clsDef";
export * from "./documentTitle";
export * from "./embed.r";
export * from "./loop";
export * from "./provide.r";
export * from "./timing";
