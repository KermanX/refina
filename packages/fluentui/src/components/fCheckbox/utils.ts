import { DArray, getDArray } from "refina";
import type { FCheckboxState } from "./fCheckbox.r";

/**
 * @returns the mixed state if any of the sources is mixed, true if any of the sources is true and none is false, false otherwise
 */
export function calcMixedCheckboxState(
  sources: DArray<boolean | FCheckboxState>,
): FCheckboxState {
  const sourcesSet = new Set(getDArray(sources));
  return sourcesSet.has("mixed")
    ? "mixed"
    : sourcesSet.has(true)
    ? sourcesSet.has(false)
      ? "mixed"
      : true
    : false;
}
