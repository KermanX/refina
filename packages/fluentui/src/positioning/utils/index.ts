export * from "./fromFloatingUIPlacement";
export * from "./mergeArrowOffset";
export * from "./parseFloatingUIPlacement";
export * from "./resolvePositioningShorthand";
export * from "./toFloatingUIPlacement";
