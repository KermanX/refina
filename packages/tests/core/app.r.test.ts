import { $app } from "refina";
import { initRootElement } from "../utils";

beforeAll(() => {
  initRootElement();
});

describe("simple app", () => {
  it("should render a simple app", async () => {
    const appInstance = $app(_ => {
      _._h1({}, "Hello World!");
      _.$cls`main`;
      _._div({}, _ => {
        _.$css`color: red`;
        _._p({}, "This is a paragraph.");
      });
    });

    await appInstance.promises.DOMUpdated;

    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
