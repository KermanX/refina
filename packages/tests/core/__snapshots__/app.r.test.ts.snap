// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`simple app > should render a simple app 1`] = `"<div id=\\"app\\"><h1>Hello World!</h1><div class=\\"main \\"><p style=\\"color: red;\\">This is a paragraph.</p></div></div>"`;
